sdk/servicebus/service-bus/src/receivers/bufferReceiver.ts

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { Receiver, ReceiverEvents, EventContext } from "rhea-promise";
+import { DefaultDataTransformer } from "@azure/core-amqp";
+
+export function messageIterator(
+  receiver: Receiver,
+  initialCredits: number
+): AsyncIterableIterator<string> & { close(): Promise<void> } {
+  const iterator = _iterator(receiver, initialCredits) as ReturnType<typeof messageIterator>;
+
+  iterator.close = async () => {
+    return new Promise((resolve) => {
+      receiver.drain = true;
+      receiver.addCredit(1);
+
+      receiver.once(ReceiverEvents.receiverDrained, () => {
+        resolve();
+      });
+    });
+  };
+
+  return iterator;
+}
+
+// TODO: first pass - infinite interator
+async function* _iterator(
+  receiver: Receiver,
+  initialCredits: number
+): AsyncIterableIterator<string> {
+  let promise = setupReceiver(receiver);
+
+  console.log(`Adding in ${initialCredits}`);
+  receiver.addCredit(initialCredits);
+
+  while (true) {
+    let message: string;
+    console.log(`Awaiting on promise for message`);
+    ({ message, promise } = await promise);
+    console.log(`Yielding next message: ${message}`);
+    yield message;
+    console.log(`Done yielding, adding 1 credit`);
+    receiver.addCredit(1);
+  }
+}
+
+interface PromisePayload {
+  promise: Promise<PromisePayload>;
+  message: string;
+}
+
+function setupReceiver(receiver: Receiver): Promise<PromisePayload> {
```

Comment:
This is an interesting way to turn an event emitter into a chain of promises. I had to stare at it a while but this seems like it'd get the job done. 

I thought about what if we had a simpler implementation where you add a single credit, then resolve once you get a message. It'd be easier to reason about but then again this solution will let us handle prefetching.

Are you planning on having a buffer for messages like we have for event hubs? (useful for prefetching messages)

Reply:
I've had a few idle thoughts around this but I want to write them down. For instance, we could make the _user_ pass in a array (user being 'us', but from a higher level in the SDK). Then we just push onto an array and then the promise just signals that data has arrived.

(you had a similar/same idea with your Queue class)